src/python/pants/backend/go/util_rules/go_mod.py

```diff
@@ -9,16 +9,58 @@
 
 from pants.backend.go.target_types import GoModSourcesField
 from pants.backend.go.util_rules.sdk import GoSdkProcess
+from pants.base.specs import AddressSpecs, AscendantAddresses
 from pants.build_graph.address import Address
 from pants.engine.engine_aware import EngineAwareParameter
 from pants.engine.fs import Digest, RemovePrefix
 from pants.engine.process import ProcessResult
 from pants.engine.rules import Get, MultiGet, collect_rules, rule
-from pants.engine.target import HydratedSources, HydrateSourcesRequest, WrappedTarget
+from pants.engine.target import (
+    HydratedSources,
+    HydrateSourcesRequest,
+    UnexpandedTargets,
+    WrappedTarget,
+)
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True)
+class OwningGoModRequest(EngineAwareParameter):
+    address: Address
+
+    def debug_hint(self) -> str:
+        return self.address.spec
+
+
+@dataclass(frozen=True)
+class OwningGoMod:
+    address: Address
+
+
+@rule
+async def find_nearest_go_mod(request: OwningGoModRequest) -> OwningGoMod:
+    # We don't expect `go_mod` targets to be generated, so we can use UnexpandedTargets.
+    candidate_targets = await Get(
+        UnexpandedTargets, AddressSpecs([AscendantAddresses(request.address.spec_path)])
+    )
+
+    # Sort by address.spec_path in descending order so the nearest go_mod target is sorted first.
+    go_mod_targets = sorted(
+        (tgt for tgt in candidate_targets if tgt.has_field(GoModSourcesField)),
+        key=lambda tgt: tgt.address.spec_path,
+        reverse=True,
+    )
+    if not go_mod_targets:
+        raise Exception(
+            f"The target {request.address} does not have any `go_mod` target in any ancestor "
```

Comment:
Given that the go.mod could be in the same directory, maybe clarify "ancestor" to "ancestor or sibling BUILD files"?